File: backend/app/services/dashboard_service.py
import logging
from collections import defaultdict

from app.domain.entities import DashboardSummary
from app.infrastructure.dynamodb.transaction_repo import TransactionRepository

logger = logging.getLogger(__name__)


class DashboardService:
    def __init__(self, transaction_repo: TransactionRepository):
        self._repo = transaction_repo

    def get_summary(
        self, user_id: str, month: int, year: int
    ) -> DashboardSummary:
        start_date = f"{year}-{month:02d}-01"
        if month == 12:
            end_date = f"{year + 1}-01-01"
        else:
            end_date = f"{year}-{month + 1:02d}-01"

        transactions = self._repo.get_by_user(
            user_id=user_id, start_date=start_date, end_date=end_date
        )

        total_spending = sum(t.amount for t in transactions)

        category_totals: dict[str, float] = defaultdict(float)
        daily_totals: dict[str, float] = defaultdict(float)

        for txn in transactions:
            category_totals[txn.category.value] += txn.amount
            daily_totals[txn.date] += txn.amount

        monthly_trend = [
            {"date": date, "amount": round(amount, 2)}
            for date, amount in sorted(daily_totals.items())
        ]

        return DashboardSummary(
            total_spending=round(total_spending, 2),
            category_breakdown={k: round(v, 2) for k, v in category_totals.items()},
            monthly_trend=monthly_trend,
            transaction_count=len(transactions),
        )

File: backend/src/services/dashboardService.js
import { getTransactionsByUser } from "../infrastructure/dynamodb.js";

export async function getDashboardSummary(userId, month, year) {
  const startDate = `${year}-${String(month).padStart(2, "0")}-01`;
  const endMonth = month === 12 ? 1 : month + 1;
  const endYear = month === 12 ? year + 1 : year;
  const endDate = `${endYear}-${String(endMonth).padStart(2, "0")}-01`;

  const transactions = await getTransactionsByUser(userId, startDate, endDate);

  let totalSpending = 0;
  const categoryTotals = {};
  const dailyTotals = {};

  for (const txn of transactions) {
    totalSpending += txn.amount;
    categoryTotals[txn.category] = (categoryTotals[txn.category] || 0) + txn.amount;
    dailyTotals[txn.date] = (dailyTotals[txn.date] || 0) + txn.amount;
  }

  const monthlyTrend = Object.entries(dailyTotals)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([date, amount]) => ({ date, amount: Math.round(amount * 100) / 100 }));

  return {
    total_spending: Math.round(totalSpending * 100) / 100,
    category_breakdown: Object.fromEntries(
      Object.entries(categoryTotals).map(([k, v]) => [k, Math.round(v * 100) / 100])
    ),
    monthly_trend: monthlyTrend,
    transaction_count: transactions.length,
  };
}
